OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/string_analysis/interpretation/AbstractStringInterpreter.scala

```diff
@@ -0,0 +1,213 @@
+                            }
+                            ep
+                    }.to[ListBuffer]
+            }.to[ListBuffer]
+    }.to[ListBuffer]
+
+    /**
+     * This function checks whether the interpretation of parameters, as, e.g., produced by
+     * [[evaluateParameters()]], is final or not and returns all refineables as a list. Hence, if
+     * this function returns an empty list, all parameters are fully evaluated.
+     */
+    protected def getNonFinalParameters(
+        evaluatedParameters: Seq[Seq[Seq[EOptionP[Entity, StringConstancyProperty]]]]
+    ): List[EOptionP[Entity, StringConstancyProperty]] =
+        evaluatedParameters.flatten.flatten.filter { _.isRefinable }.toList
+
+    /**
+     * convertEvaluatedParameters takes a list of evaluated / interpreted parameters as, e.g.,
+     * produced by [[evaluateParameters]] and transforms these into a list of lists where the inner
+     * lists are the reduced [[StringConstancyInformation]]. Note that this function assumes that
+     * all results in the inner-most sequence are final!
+     */
+    protected def convertEvaluatedParameters(
+        evaluatedParameters: Seq[Seq[Seq[EOptionP[Entity, Property]]]]
+    ): ListBuffer[ListBuffer[StringConstancyInformation]] = evaluatedParameters.map { paramList ⇒
+        paramList.map { param ⇒
+            StringConstancyInformation.reduceMultiple(
+                param.map {
+                    _.asFinal.p.asInstanceOf[StringConstancyProperty].stringConstancyInformation
+                }
+            )
+        }.to[ListBuffer]
+    }.to[ListBuffer]
+
+    /**
+     *
+     * @param instr The instruction that is to be interpreted. It is the responsibility of
+     *              implementations to make sure that an instruction is properly and comprehensively
+     *              evaluated.
+     * @param defSite The definition site that corresponds to the given instruction. `defSite` is
+     *                not necessary for processing `instr`, however, may be used, e.g., for
+     *                housekeeping purposes. Thus, concrete implementations should indicate whether
+     *                this value is of importance for (further) processing.
+     * @return The interpreted instruction. A neutral StringConstancyProperty contained in the
+     *         result indicates that an instruction was not / could not be interpreted (e.g.,
+     *         because it is not supported or it was processed before).
+     *         <p>
+     *         As demanded by [[InterpretationHandler]], the entity of the result should be the
+     *         definition site. However, as interpreters know the instruction to interpret but not
+     *         the definition site, this function returns the interpreted instruction as entity.
+     *         Thus, the entity needs to be replaced by the calling client.
+     */
+    def interpret(instr: T, defSite: Int): EOptionP[Entity, Property]
```

Comment:
can the return type be specified more precisely?

Reply:
In some cases yes. Fixed now.